### PR TITLE
Validate connection manager/inbound governor transition order

### PR DESCRIPTION
**Pull request by bolt12**

### CODE REVIEW

File: ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Core.hs [hunk 1]
```diff
@@ -1187,13 +1191,13 @@ withConnectionManager ConnectionManagerArguments {
...
                         modifyTMVarPure_ stateVar $
                           (Map.update (\connVar' ->
-                            if eqTVar (Proxy :: Proxy m) connVar' connVar
-                               then Nothing
-                               else Just connVar')
-                          peerAddr)
+                                      if eqTVar (Proxy :: Proxy m) connVar' connVar
+                                          then Nothing
+                                          else Just connVar')
+                                      peerAddr)
```
Comment (coot) [inline ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Core.hs]: This change does not belong to this patch. I also prefer the previous one since it fits in 80 columns.

File: ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Core.hs [hunk 2]
```diff
@@ -530,359 +549,417 @@ withConnectionManager ConnectionManagerArguments {
...
+                  tr = TransitionTrace peerAddr (mkTransition connState connState')
+                  absConnState = abstractState (Known connState)
+                  shouldTrace = absConnState /= TerminatedSt
+
+              writeTVar connVar connState'
+              return (connState, tr, shouldTrace)
+
+            when shouldTrace $
+              traceWith trTracer tr
+            traverse_ cancel (getConnThread connState)
```
[the remark this answers is not in the document]
Comment (bolt12) [reply ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Core.hs:589]: I have experimented with using `modifyTMVar` and it seems it can deadlock so maybe we are better off with this solution.

File: ouroboros-network-framework/test/Test/Ouroboros/Network/Server2.hs [hunk 3]
```diff
@@ -1933,58 +1936,102 @@ data Three a b c
...
+    sim = multiNodeSim serverAcc dataFlow
+                       (Script (toBearerInfo absBi :| [noAttenuation]))
+                       maxAcceptedConnectionsLimit l
+
+
+-- | Property wrapping `multinodeExperiment`.
+--
+-- Note: this test validates inbound governor state changes.
+--
+prop_multinode_ig_Sim :: Int
```
[the remark this answers is not in the document]
Comment (coot) [reply ouroboros-network-framework/test/Test/Ouroboros/Network/Server2.hs]: Either sounds good to me.

File: ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Core.hs [hunk 4]
```diff
@@ -2060,6 +2220,13 @@ modifyTMVar v k =
     atomically (putTMVar v a')
     return b
 
+modifyTMVarSTM :: MonadSTM m => StrictTMVar m a -> (a -> STM m (a, b)) -> STM m b
+modifyTMVarSTM v k = do
+    a <- takeTMVar v
+    (!a',b) <- k a
```
Comment (coot) [inline ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Core.hs:2226]:
```suggestion
    (a', b) <- k a
```
no need for the bang since we use `StrictTVar`.  It's also not needed in other `stm` primitives in this module.  Either a github issue to remove them should be enough.

**Review by coot — APPROVED** [review]
LGTM; I left a minor comment (bang in `modifyTMVarSTM`), but this can be left for later as well (in that case create a github issue for it).
